drivers/power/ltc3337/ltc3337.c

```diff
@@ -0,0 +1,659 @@
+#define LTC3337_RC_ADC_READY_POS    4
+#define LTC3337_RC_ADC_READY        ((uint16_t)1 << LTC3337_RC_ADC_READY_POS)
+#define LTC3337_RC_IPK_PIN_MSK      0x7
+#define LTC3337_RC_IPK_PIN_POS      5
+#define LTC3337_RC_DIE_TEMP_MSK     0xFF
+#define LTC3337_RC_DIE_TEMP_POS     8
+
+#define LTC3337_REG_D   0x04 //BAT_IN V, Ipeak On
+#define LTC3337_REG_E   0x05 //BAT_IN V, Ipeak Off
+#define LTC3337_REG_F   0x06 //BAT_OUT V, Ipeak On
+#define LTC3337_REG_G   0x07 //BAT_OUT V, Ipeak Off
+#define LTC3337_BATV_MSK            0xFFF
+
+#define LTC3337_REG_H   0x08 //Temp Alarm Config
+#define LTC3337_RH_TEMP_MSK         0xFF
+#define LTC3337_RH_HOT_ALARM_POS    8
+#define LTC3337_RH_COLD_ALARM_POS   0
+
+
+#define LTC3337_VLSB_uV     1465 //1.465 mV = 1465 uV
+#define LTC3337_TLSB_mC     784  //0.784 Deg C = 784 mC
+#define LTC3337_TEMP_MIN_C  -41
+
+#define LTC3337_NUM_IPEAKS  8 //3-bit iPeak Input
+
+//Scale values used in integer match to calulate A/nA-hrs from counters
+#define LTC3337_NANO_AMP        1000000000
+#define LTC3337_CALC_SCALE      10000
+#define LTC3337_CALC_TO_WHOLE   (LTC3337_NANO_AMP / LTC3337_CALC_SCALE)
+
+/**
+ * Structure supporting the iPeak lookup table
+ */
+struct ipeak_entry_t {
+	uint8_t  iPeak_mA;       //iPeak, in mA
+	uint32_t qLsb_nA_hr;     //qLsb, nA-hr component
+};
+
+/**
+ * Lookup table for the iPeak values. This converts the 3-bit
+ * iPeak input configuration signals, to a iPeak current and qLsb
+ * values for Coulomb counter calculations
+ */
+static const struct ipeak_entry_t iPeakLookupTable[LTC3337_NUM_IPEAKS] = {
+	{ .iPeak_mA = 5,   .qLsb_nA_hr = 745700 },   //b000 - 5ma, 745.7 uA-hr
+	{ .iPeak_mA = 10,  .qLsb_nA_hr = 1491000 },  //b001 - 10ma, 1.491 mA-hr
+	{ .iPeak_mA = 15,  .qLsb_nA_hr = 2237000 },  //b010 - 15ma, 2.237 mA-hr
+	{ .iPeak_mA = 20,  .qLsb_nA_hr = 2983000 },  //b011 - 20ma, 2.983 mA-hr
+	{ .iPeak_mA = 25,  .qLsb_nA_hr = 3728000 },  //b100 - 25ma, 3.728 mA-hr
+	{ .iPeak_mA = 50,  .qLsb_nA_hr = 7457000 },  //b101 - 50ma, 7.457 mA-hr
+	{ .iPeak_mA = 75,  .qLsb_nA_hr = 11180000 }, //b110 - 75ma, 11.18 mA-hr
+	{ .iPeak_mA = 100, .qLsb_nA_hr = 14910000 }, //b111 - 100ma, 14.91 mA-hr
+};
+
+/** Local Prototypes */
+static int  ltc3337_read_reg(struct ltc3337_dev* dev, uint8_t reg,
+			     uint16_t* data);
+static int  ltc3337_write_reg(struct ltc3337_dev* dev, uint8_t reg,
+			      uint16_t data);
+static uint32_t ltc3337_vbat_to_mV(uint16_t rawValue);
+static int16_t  ltc3337_temp_reg_to_C(uint8_t rawValue);
+static uint8_t  ltc3337_temp_C_to_reg(int16_t tempC);
+static void     ltc3337_charge_count_to_A(uint16_t chargeCount,
+		uint8_t prescale,
+		uint32_t lsbnA,       struct charge_count_t* value);
+static uint16_t ltc3337_A_to_charge_count(uint8_t prescale, uint32_t lsbnA,
+		struct charge_count_t* chargeA);
+
+
+
+int ltc3337_init(struct ltc3337_dev** device,
+		 struct ltc3337_init_param* init_param)
+{
+	int32_t ret;
```

Comment:
Function return type is int, but ret variable is defined as int32_t. It should be int.